### DISCUSSION TITLE
Is it possible to select ASIO devices?

### DISCUSSION THREAD

**Discussion by jonblk** [FluidSynth/fluidsynth#1340] [Q&A] [answered]
Is Fluidsynth able to use Asio devices such as FlexAsio and ASIO4ALL?
If so, how can I list the available Asio devices, and how can I set it?

(i am looking to set it via command line like this:  `fluidsynth -a wasapi  -o audio.wasapi.device=ASIO4ALL xxx.sf2)`

Thank you!

**Comment by derselbst** [comment, marked as answer]
Pls. execute `fluidsynth --query-audio-devices` to see which audio devices are available via the wasapi driver. If it's not listed, you're out of luck.

**Reply by jonblk** [reply]
It appears that ASIO devices are not listed. 

Thank you for the response.

**Comment by PeterHevesi** [comment]
Would it be possible, that you add support for ASIO devices?
It would be super cool :)

**Comment by ReinholdH** [comment]
ASIO can be added to fluidsynth through Portaudio. See [here](https://www.fluidsynth.org/api/CreatingAudioDriver.html)
Portaudio offers ASIO support. See [here](https://www.portaudio.com/docs/v19-doxydocs/compile_windows_asio_msvc.html) for Windows.
Please be aware of the license agreement of ASIO. For commercial apps which use fluidsynth under LGPL and Portaudio under the free [Portaudio license](https://www.portaudio.com/license.html)   a special contract needs to be signed with Steinberg.

**Comment by PeterHevesi** [comment]
@ReinholdH, thanks so much for this tip :)
But can I also add support for Asio4All through this Portaudio?
Because I am actually not interested in standard ASIO, but rather Asio4All, since many people don't have ASIO sound cards, and only have sound card integrated to motherboard. With Asio4All, you can achieve fantastic latency (5ms when using 128 buffer size, and 12 ms when using 512 buffer size) and with standard soundcard integrated on the motherboard that all PCs have :)

**Comment by ReinholdH** [comment]
I do not know if Portaudio supports Asio4All. We use the software in a commercial app. For us Asio4All is not an option due to 

- our customer community where the installation of an additional Asio4All driver is not manageable
- licensing of Asio4All does not allow a commercial bundling with an app ([see](https://asio4all.org/about/))

All I can say here is that libfluidsynth + Portaudio + ASIO works brilliant with regards to latency. 
For questions regarding Portaudio please ask the Portaudio community.

**Reply by PeterHevesi** [reply]
Thanks for the info :) I have wrote the guy that created ASIO4ALL (Michael Tippach) and asked him, if we can automatically install ASIO4ALL for our customers. Can't see why he wouldn't allow me to do it, since ASIO4ALL is free, and Michael don't even accept donations :D

**Comment by syntaro** [comment]
Well I don't know good place for discussion about problem on ASIO4All and portaudio.
When portaudio enumerate ASIO4All it scan and open connection wasapi and asio. So two times..it will not work. You(thread owner) need customize portaudio code.

**Comment by jimhen3ry** [comment]
My understanding is that ASIO4ALL is a wrapper for WASAPI on Windows. The intent seems to be to provide a way to use WASAPI on systems that don't have ASIO hardware from apps that only provide an ASIO option. I think you will get equally good results from FluidSynth on Windows by using the WASAPI driver.

**Comment by syntaro** [comment]
Excuse me.
My (you) meant (thread owner).
Well I did make app doing ASIO4All + PortAudio  before.

And Im thinking WASAPI is good enough for general use..because don't need tuning on too many hardware.
